R/aaa_quantiles.R

```diff
@@ -21,23 +21,100 @@ check_quantile_level <- function(x, object, call) {
   x
 }
 
-# Assumes the columns have the same order as quantile_level
-restructure_rq_pred <- function(x, object) {
-  num_quantiles <- NCOL(x)
-  if ( num_quantiles == 1L ){
-    x <- matrix(x, ncol = 1)
+
+# -------------------------------------------------------------------------
+# A column vector of quantiles with an attribute
+
+#' @export
+vec_ptype_abbr.vctrs_quantiles <- function(x, ...) "qntls"
+
+#' @export
+vec_ptype_full.vctrs_quantiles <- function(x, ...) "quantiles"
+
+#' @importFrom rlang is_double !!!
+new_vec_quantiles <- function(values = list(), quantile_levels = double()) {
+  quantile_levels <- vctrs::vec_cast(quantile_levels, double())
+  vctrs::new_vctr(
+    values, quantile_levels = quantile_levels, class = "vctrs_quantiles"
+  )
+}
+
+
+#' Create a vector containing sets of quantiles
+#'
+#' @param values A matrix of values. Each column should correspond to one of
+#'   the quantile levels.
+#' @param quantile_levels A vector of probabilities corresponding to `values`.
+#'
+#' @export
+#' @return A vector of values associated with the quantile levels.
+#'
+#' @examples
+#' v <- vec_quantiles(matrix(rnorm(20), 5), c(.2, .4, .6, .8))
+#'
+#' # Access the underlying information
+#' attr(v, "quantile_levels")
+#' vctrs::vec_data(v)
+vec_quantiles <- function(values, quantile_levels = double()) {
+  check_vec_quantiles_inputs(values, quantile_levels)
+  quantile_levels <- vctrs::vec_cast(quantile_levels, double())
+  num_lvls <- length(quantile_levels)
+
+  if (ncol(values) != num_lvls) {
+    cli::cli_abort(
+      "The number of columns in {.arg values} must be equal to the length of
+        {.arg quantile_levels}."
+    )
   }
-  n <- nrow(x)
+  values <- lapply(vctrs::vec_chop(values), drop)
+  new_vec_quantiles(values, quantile_levels)
+}
 
+check_vec_quantiles_inputs <- function(values, levels) {
```

Comment:
```suggestion
check_vec_quantiles_inputs <- function(values, levels, call = caller_env()) {
```

Then, in each of the calls to `cli::cli_abort()` in this function, pass `call = call` so that users see "Error in `user_facing_fn()`" rather than "Error in `unexported_helper_fn()`"

Reply:
Note that this doesn't apply to `vec_quantiles()`, as the default `caller_env()` refers to the correct function in that case!